### PR TITLE
option to use customized client factory

### PR DESCRIPTION
**Pull request by yiminc**
**What changed?**
Add ServerOption to use customized ClientFactory

**Why?**
This will make it possible to override client factory to have different customized logic.

**How did you test it?**
Existing integration tests

**Potential risks**
No.

**Is hotfix candidate?**
No.

### CODE REVIEW

File: client/clientfactory.go [hunk 1]
```diff
@@ -60,6 +59,18 @@ type (
 		NewAdminClientWithTimeout(rpcAddress string, timeout time.Duration, largeTimeout time.Duration) (adminservice.AdminServiceClient, error)
 	}
 
+	// FactoryProvider can be used to provide a customized client Factory implementation.
+	FactoryProvider interface {
+		NewFactory(
+			rpcFactory common.RPCFactory,
+			monitor membership.Monitor,
+			metricsClient metrics.Client,
+			dc *dynamicconfig.Collection,
```
Comment (wxing1292) [inline client/clientfactory.go:68]: should this be an interface?

File: temporal/server_options.go [hunk 2]
```diff
@@ -59,7 +60,8 @@ type (
 		persistenceServiceResolver resolver.ServiceResolver
 		elasticseachHttpClient     *http.Client
 		dynamicConfigClient        dynamicconfig.Client
-		customDataStoreFactory     client.AbstractDataStoreFactory
+		customDataStoreFactory     persistenceClient.AbstractDataStoreFactory
+		clientFactoryProvider      client.FactoryProvider
```
Comment (wxing1292) [inline temporal/server_options.go:64]: customClientFactory?

**Review by wxing1292 — APPROVED** [review]

File: client/clientfactory.go (hunk 1 above)
Comment (yiminc) [reply client/clientfactory.go:68]: Yes, so this would make it possible to supply customized factory.

File: temporal/server_options.go (hunk 2 above)
Comment (yiminc) [reply temporal/server_options.go:64]: This is not factory itself, this is factory provider that provide a factory.

**Pull request merged**